components/user-service/src/main/java/edu/jhuapl/dorset/ldapuserservice/LdapUserService.java

```diff
@@ -0,0 +1,198 @@
+        for (int i = 0; i < userAttributes.length; i++) {
+            userAttributes[i] = userAttributes[i].trim();
+        }
+        this.setUserAttributes(userAttributes);
+  
+        Hashtable<String, String> env = new Hashtable<String, String>();
+        env.put(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.ldap.LdapCtxFactory");
+        env.put(Context.SECURITY_AUTHENTICATION, "simple");
+
+        env.put(Context.PROVIDER_URL, this.ldapServer);
+        env.put(Context.SECURITY_PRINCIPAL, this.ldapBindDn);
+        env.put(Context.SECURITY_CREDENTIALS, this.ldapPassword);
+
+        try {
+            this.ctx = new InitialDirContext(env);
+        } catch (NamingException ex) {
+            logger.info("Error when trying to establish the context.");
+        }    
+        
+        this.setContext(this.ctx);
+            
+    }
+
+    @Override
+    public String create(User user) throws UserException {
+        throw new UnsupportedOperationException("LdapUserService does not implement Create.");
+    }
+
+    @Override
+    public void delete(String userName) {
+        throw new UnsupportedOperationException("LdapUserService does not implement Delete.");
+    }
+
+    @Override
+    public User getUser(String userName) {
+        User user = null;
+
+        // check if user exists in the hash map
+        user = this.users.get(userName);
+
+        if (user == null) {
+
+            Properties props = new Properties();
+            props.setProperty("userName", userName);
+
+            try {
+                this.retrieve(props);
+            } catch (UserException e) {
+                logger.debug("User not found: " + userName);
+                return null;
+            }
+            user = this.users.get(userName);
+        }
+
+        return user;
+    }
+    
+    @Override
+    public String retrieve(Properties properties) throws UserException {
+        User user = new User();
+        String userName = properties.getProperty("userName");
+
+        String key;
+        Attribute value;
+        
+        SearchControls searchControls = getSearchControls(this.userAttributes);
+        ctx = getContext();
+
+        try {
+            NamingEnumeration<SearchResult> answer = ctx.search(this.ldapSearchBase,
+                            this.ldapFilterAttribute + "=" + userName, searchControls);
+            if (answer.hasMore()) {
+                Attributes attrs = answer.next().getAttributes();
+                NamingEnumeration<String> attrsKeys = attrs.getIDs();
+                while (attrsKeys.hasMoreElements()) {
+                    key = attrsKeys.next();
+                    value = attrs.get(key);
+                    user.setUserInformation(key, value.toString());
+                }
+            } else { 
+                logger.info("user not found: " + userName);
+                throw new UserException("User not found: " + userName + ". ");
+            }
```

Comment:
It is generally a bad idea to both log an error message and throw an exception. This is because the calling component gets the exception and has the opportunity to handle it or log it or so on. If it handles the exception, then you have an error message when none is needed. If it logs the error, you now have 2 log messages (and I've seen extreme cases where many components logged and rethrew the exception so that you ended up with many log messages about a single error).

I think we should just throw the exception and let the calling component decide whether to log or not.